lib/spack/spack/cmd/view.py

```diff
@@ -48,258 +48,187 @@
-
-    dotspack = transform_path(spec, '.spack', path)
-    if os.path.exists(dotspack):
-        tty.warn('Skipping existing package: "%s"' % spec.name)
-        return
-
-    if verbose:
-        tty.info('Linking package: "%s"' % spec.name)
-    for dirpath, dirnames, filenames in os.walk(spec.prefix):
-        if not filenames:
-            continue        # avoid explicitly making empty dirs
-
-        targdir = transform_path(spec, dirpath, path)
-        assuredir(targdir)
-
-        for fname in filenames:
-            src = os.path.join(dirpath, fname)
-            dst = os.path.join(targdir, fname)
-            if os.path.exists(dst):
-                if '.spack' in dst.split(os.path.sep):
-                    continue    # silence these
-                tty.warn("Skipping existing file: %s" % dst)
-                continue
-            link(src, dst)
-
-
-def visitor_symlink(specs, args):
-    'Symlink all files found in specs'
-    path = args.path[0]
-    assuredir(path)
-    for spec in specs:
-        link_one(spec, path, verbose=args.verbose)
-
-
-visitor_add = visitor_symlink
-visitor_soft = visitor_symlink
-
-
-def visitor_hardlink(specs, args):
-    'Hardlink all files found in specs'
-    path = args.path[0]
-    assuredir(path)
-    for spec in specs:
-        link_one(spec, path, os.link, verbose=args.verbose)
-
-
-visitor_hard = visitor_hardlink
-
-
-def visitor_remove(specs, args):
-    'Remove all files and directories found in specs from args.path'
-    path = args.path[0]
-    for spec in specs:
-        remove_one(spec, path, verbose=args.verbose)
-    purge_empty_directories(path)
-
-
-visitor_rm = visitor_remove
-
+def view(parser, args):
+    'Produce a view of a set of packages.'
 
-def visitor_statlink(specs, args):
-    'Give status of view in args.path relative to specs'
     path = args.path[0]
-    for spec in specs:
-        check_one(spec, path, verbose=args.verbose)
 
+    view = YamlFilesystemView(
+        path, spack.store.layout,
+        ignore_conflicts=getattr(args, "ignore_conflicts", False),
+        link=os.hardlink if args.action in ["hardlink", "hard"]
+        else os.symlink,
+        verbose=args.verbose)
+
+    # Process common args and specs
+    if getattr(args, "all", False):
+        specs = view.get_all_specs()
+        if len(specs) == 0:
+            tty.warn("Found no specs in %s" % path)
+
+    elif args.action in actions_link:
+        # only link commands need to disambiguate specs
+        specs = [spack.cmd.disambiguate_spec(s) for s in args.specs]
+
+    elif args.action in actions_status:
+        # no specs implies all
+        if len(args.specs) == 0:
+            specs = view.get_all_specs()
+        else:
+            specs = relaxed_disambiguate(args.specs, view)
+
+    else:
+        # status and remove can map the name to packages in view
+        specs = relaxed_disambiguate(args.specs, view)
+
+    activated = list(filter(lambda s: s.package.is_extension and
```

Comment:
The main goal of collecting these is to avoid a case where an extendee package has globally activated extensions and we add it to the view. I think a stronger check here would be to grab the extendee package and check `activated_extensions_for`. This would also mean changing the checks in `test_view_extension_global_activation`.